theano/tensor/opt.py

```diff
@@ -221,89 +221,124 @@ def inplace_elemwise_optimizer(fgraph):
-                                if i not in baseline.values() \
-                                    and not isinstance(node.inputs[i],
-                                                       Constant)\
-                                    and not fgraph.destroyers(node.inputs[i])\
-                                    and node.inputs[i] not in protected_inputs]
-
-            verbose = False
-
-            raised_warning = not verbose
-
-            for candidate_output in candidate_outputs:
-                for candidate_input in candidate_inputs:
-                    # remove inputs that don't have the same dtype as the output
-                    if node.inputs[candidate_input].type != node.outputs[
-                        candidate_output].type:
-                        continue
-
-                    inplace_pattern = dict(baseline)
-                    inplace_pattern[candidate_output] = candidate_input
-                    try:
-                        if hasattr(op.scalar_op, "make_new_inplace"):
-                            new_scal = op.scalar_op.make_new_inplace(
-                                scalar.transfer_type(
-                                    *[inplace_pattern.get(i, None) \
-                                          for i in xrange(len(node.outputs))]))
-                        else:
-                            new_scal = op.scalar_op.__class__(
-                                scalar.transfer_type(
-                                    *[inplace_pattern.get(i, None) \
-                                          for i in xrange(len(node.outputs))]))
-                        new_outputs = OP(new_scal, inplace_pattern)(
+        # gpuarray GpuElemwise inherit from Elemwise
+        elemwise_nodelist = [node for node in
+                             list(graph.io_toposort(fgraph.inputs,
+                                                    fgraph.outputs))
+                             if type(node.op) == OP]
+
+        # See: https://github.com/Theano/Theano/issues/2962
+        if len(elemwise_nodelist) > 0:
+            check_each_change = int(numpy.ceil(numpy.log(
+                len(elemwise_nodelist))))
+            check_each_change = max(check_each_change, 1)
+        else:
+            return
+        failed_list = []
+        prev_len_elemlist = len(elemwise_nodelist) + 1
+        changed_node_since_validate = []
+        while check_each_change > 0 and prev_len_elemlist > len(elemwise_nodelist):
+            prev_len_elemlist = len(elemwise_nodelist)
+            for node in elemwise_nodelist:
+                op = node.op
+                baseline = op.inplace_pattern
+                protected_inputs = [
+                    f.protected for f in node.fgraph._features if
+                    isinstance(f, theano.compile.function_module.Supervisor)]
+                protected_inputs = sum(protected_inputs, [])  # flatten the list
+                protected_inputs.extend(fgraph.outputs)
```

Comment:
The old version was like that, but I think we can move the protected_inputs computation out of the while loop. Just a small speed up.